src/modules/commander/Commander.cpp

```diff
@@ -1790,92 +1791,148 @@ void Commander::executeActionRequest(const action_request_s &action_request)
+
+			get_distance_to_point_global_wgs84(_home_pub.get().lat, _home_pub.get().lon, _home_pub.get().alt,
+							   lat, lon, alt,
+							   &home_dist_xy, &home_dist_z);
+
+			eph = gps.eph;
+			epv = gps.epv;
+		}
+	}
+
+	return (home_dist_xy > eph * 2.f) || (home_dist_z > epv * 2.f);
+}
+
 /**
 * @brief This function initializes the home position an altitude of the vehicle. This happens first time we get a good GPS fix and each
 *		 time the vehicle is armed with a good GPS fix.
 **/
 bool
 Commander::set_home_position()
 {
-	// Need global and local position fix to be able to set home
-	// but already set the home position in local coordinates if available
-	// in case the global position is only valid after takeoff
-	if (_param_com_home_en.get() && _status_flags.local_position_valid) {
+	bool updated = false;
+	home_position_s home{};
 
+	if (_status_flags.local_position_valid) {
 		// Set home position in local coordinates
 		const vehicle_local_position_s &lpos = _local_position_sub.get();
-		_heading_reset_counter = lpos.heading_reset_counter;
+		_heading_reset_counter = lpos.heading_reset_counter; // TODO: should not be here
 
-		home_position_s home{};
-		home.timestamp = hrt_absolute_time();
-		home.manual_home = false;
 		fillLocalHomePos(home, lpos);
+		updated = true;
+	}
 
-		if (_status_flags.global_position_valid) {
-
-			const vehicle_global_position_s &gpos = _global_position_sub.get();
+	if (_status_flags.global_position_valid) {
+		// Set home using the global position estimate (fused INS/GNSS)
+		const vehicle_global_position_s &gpos = _global_position_sub.get();
+		fillGlobalHomePos(home, gpos);
+		setHomePosValid();
+		updated = true;
+
+	} else if (_status_flags.gps_position_valid) {
+		// Set home using GNSS position
+		vehicle_gps_position_s gps_pos;
+		_vehicle_gps_position_sub.copy(&gps_pos);
+		const double lat = static_cast<double>(gps_pos.lat) * 1e-7;
+		const double lon = static_cast<double>(gps_pos.lon) * 1e-7;
+		const float alt = static_cast<float>(gps_pos.alt) * 1e-3f;
+		fillGlobalHomePos(home, lat, lon, alt);
+		setHomePosValid();
+		updated = true;
+
+	} else if (_local_position_sub.get().z_global) {
+		// handle special case where we are setting only altitude using local position reference
+		// This might be overwritten by altitude from global or GNSS altitude
+		home.timestamp = hrt_absolute_time();
+		home.alt = _local_position_sub.get().ref_alt;
+		home.valid_alt = true;
 
-			// Ensure that the GPS accuracy is good enough for intializing home
-			if (isGPosGoodForInitializingHomePos(gpos)) {
-				fillGlobalHomePos(home, gpos);
-				setHomePosValid();
-			}
-		}
+		updated = true;
+	}
 
-		_home_pub.update(home);
+	if (updated) {
+		home.timestamp = hrt_absolute_time();
+		home.manual_home = false;
+		updated = _home_pub.update(home);
 	}
 
-	return _status_flags.home_position_valid;
+	return updated;
 }
 
 bool
 Commander::set_in_air_home_position()
 {
-	if (_param_com_home_en.get()
```

Comment:
Does `COM_HOME_EN` 0 imply `COM_HOME_IN_AIR` 0 ? The second parameter just has no effect anymore right?

Reply:
Yes, `COM_HOME_IN_AIR` has no effect when `COM_HOME_EN` = 0. We could merge those parameters: 0: don't set home, 1. set home on ground only, 2. set home on ground and in air

Reply:
That's a good idea, easier to configure. No need to do it in this pr.